qiime2/sdk/actiongraph.py

```diff
@@ -0,0 +1,215 @@
+                req.append([k, v.qiime_type])
+            else:
+                non_req.append(["."+k, [v.qiime_type, None]])
+
+    return req, non_req
+
+
+def unravel(l):
+    """
+    Unravel Union node to get all permutations of types for each action
+
+    Parameters
+    ----------
+    list : list of Qiime2.types
+
+    Returns
+    -------
+    list of lists - list of permuations of types for each action
+
+    """
+
+    result = [l]
+    for i, x in enumerate(l):
+        if len(list(x[1])) > 1:
+            members = list(x[1])
+            temp = copy.deepcopy(result)
+
+            # update result with first element of types in member
+            for each_list in result:
+                each_list[i][1] = members[0]
+
+            # add in other permutations of types in member
+            for n in range(1, len(members)):
+                copy_result = copy.deepcopy(temp)
+                for each_list in copy_result:
+                    each_list[i][1] = members[n]
+                result += copy_result
+    return result
+
+
+def generate_nodes_by_action(action, opt):
+    """
+    Given a method, generates all combinations of inputs and
+    outputs for that particular method and and stores the combinations
+    as dictionaries in a resulting list.
+
+    Parameters
+    ----------
+    method : Qiime2.action
+    opt : {True, False}
+        Delineates if optional types should be included
+
+    Returns
+    -------
+    list of dictionaries - each dictionary is a combination inputs and
+        outputs for particular node
+
+    """
+
+    input, input_nr = get_next_arguments(action, "input")
+    param, param_nr = get_next_arguments(action, "param")
+    output, output_nr = get_next_arguments(action, "output")
+
+    input = unravel(input)
+    param = unravel(param)
+
+    opt_in_list = []
+
+    if opt:
+        opt_in_list += input_nr
+        opt_in_list += param_nr
+        opt_in_list = unravel(opt_in_list)
+        ins = [dict(x) for x in
+               [list(chain.from_iterable(i)) for i in
+                list(product(input, param, opt_in_list))]]
+        outs = dict(output + output_nr)
+        results = [{'inputs': i, 'outputs': outs} for i in ins]
+        return results
+
+    ins = [dict(x) for x in
+           [list(chain.from_iterable(i)) for i in list(product(input, param))]]
+    outs = dict(output)
+    results = [{'inputs': i, 'outputs': outs} for i in ins]
+    return results
+
+
+def build_graph(sigs=[], opt=False):
+    """
+    Constructs a networkx graph with different semantic types
+    and actions as nodes
+
+    Parameters
+    ----------
+    sigs : list of Qiime2.action
+    opt : {True, False}
+        Delineates if optional types should be included in the graph
+
+    Returns
+    -------
+    nx.DiGraph - networkx graph connected based on all or specified methods
+
+    """
+
+    G = nx.DiGraph()
+    G.edges(data=True)
+    action_list = []
+    sigs = [x.__name__ for x in sigs]
```

Comment:
This line actually breaks when you provide a list of actions/methods. (see comment below for solution)